### ISSUE TITLE
Check rules serially instead of in parallel

### ISSUE THREAD

**Issue by robross0606** [CacheControl/json-rules-engine#271]
We're running into issues where our facts are based on transactional access to a database.  When we try to properly decompose our rules, we hit problems where the same transaction is trying to be used simultaneously because the rules engine appears to process rules in parallel.  Is there no way to check rules serially (on-by-one) instead of having them all checked at once?

**Comment by robross0606** [comment]
Ideally there would be a way to tell the engine to process rule conditions in order, either by setting priority on each condition, or simply by telling the engine "all in order".

**Comment by robross0606** [comment]
Another possibility would be to add "before" operators or something like that because fact comparisons _also_ force facts to be run in order instead of in parallel.

**Comment by StingyJack** [comment]
a "depends on" mapping may be the easiest and would allow for all serial, all parallel, or a mix of both.

**Comment by robross0606** [comment]
@StingyJack Is there a good example of this somewhere?

**Comment by StingyJack** [comment]
nuget / npm dependencies.  If a condition depends on another condition, then those must be done in order (serially).  The rest can be parallel processed

**Comment by robross0606** [comment]
@StingyJack are you suggesting an alternate enhancement or suggesting this feature already exists in the library?

**Comment by StingyJack** [comment]
I'm suggesting a way that we are probably going to have to implement. Unfortunately this will be in a fork and not here because we can't wait an indefinite time for the PR to be approved or risk that it doesn't. Per current company policy that will be a private fork (for now, that can change)

**Comment by robross0606** [comment]
> this will be in a fork and not here

Has any work been done on this anywhere yet?

**Comment by StingyJack** [comment]
After a bit more experience and research with this and other rules engines, when I read the original post now I see what looks like a rather large red flag due with the mention of "transaction".  Since transactions are usually necessary for when you need to change data and the purpose of a rule engine is to provide knowledge or information (but not to actually perform the actions). how does the one affect or interfere with the other?

The steps could be something like 
- some service gets fact data from the db, opening transaction if there is potential for modification
- service sends fact data for rule evaluation and gets instruction of what to do for that data
- service executes action, including either commit or rollback of transaction

Are you doing something like that or something else?

**Comment by robross0606** [comment]
No, this isn't changing data as you would think it does.  But it does include a "lazy copy" mechanism that assigns unique identifiers on-the-fly, which requires transactions.  The domain data is stable.  The bottom line is the same.  We need to evaluate rule facts serially instead of in parallel.

**Comment by robross0606** [comment]
> The steps could be something like
> - some service gets fact data from the db, opening transaction if there is potential for modification
> - service sends fact data for rule evaluation and gets instruction of what to do for that data
> - service executes action, including either commit or rollback of transaction

This is something like what we do already.  The problem is that the effort around "some service gets fact data" effectively replaces the point of using `json-rules-engine` to begin with.  We cannot decompose facts properly and compose them into different rules if we cannot specify that some facts must be retrieved serially instead of all being done in parallel.  That means basically collapsing each mixture of facts into a single fact with lots of code.

At that point, what roles does json-rules-engine really play?
